[package] Add an intern keyword

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57341 [package] Add an intern keyword**

Require that users be explicit about what they are going to be
interning. There are a lot of changes that are enabled by this. The new
overall scheme is:

PackageExporter maintains a dependency graph. Users can add to it,
either explicitly (by issuing a `save_*` call) or explicitly (through
dependency resolution). Users can also specify what action to take when
PackageExporter encounters a module (deny, intern, mock, extern).

Nothing (except pickles, tho that can be changed with a small amount
of work) is written to the zip archive until we are finalizing the
package. At that point, we consult the dependency graph and write out
the package exactly as it tells us to.

This accomplishes two things:
1. We can gather up *all* packaging errors instead of showing them one at a time.
2. We require that users be explicit about what's going in packages, which is a common request.

Differential Revision: [D28114185](https://our.internmc.facebook.com/intern/diff/D28114185/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28114185/)!


Differential Revision: [D28114185](https://our.internmc.facebook.com/intern/diff/D28114185)

suo committed:
tmp intern

[ghstack-poisoned]

suo committed:
Update on "tmp intern"




Differential Revision: [D28114185](https://our.internmc.facebook.com/intern/diff/D28114185)

[ghstack-poisoned]

suo committed:
Update on "tmp intern"




Differential Revision: [D28114185](https://our.internmc.facebook.com/intern/diff/D28114185)

[ghstack-poisoned]

suo committed:
Update on "tmp intern"




Differential Revision: [D28114185](https://our.internmc.facebook.com/intern/diff/D28114185/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28114185/)!


Differential Revision: [D28114185](https://our.internmc.facebook.com/intern/diff/D28114185)

[ghstack-poisoned]

suo committed:
Update on "tmp intern"




Differential Revision: [D28114185](https://our.internmc.facebook.com/intern/diff/D28114185/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28114185/)!


Differential Revision: [D28114185](https://our.internmc.facebook.com/intern/diff/D28114185)

[ghstack-poisoned]

suo committed:
Update on "tmp intern"




Differential Revision: [D28114185](https://our.internmc.facebook.com/intern/diff/D28114185/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28114185/)!


Differential Revision: [D28114185](https://our.internmc.facebook.com/intern/diff/D28114185)

[ghstack-poisoned]

suo commented:
@suo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28114185).

suo committed:
Update on "tmp intern"




Differential Revision: [D28114185](https://our.internmc.facebook.com/intern/diff/D28114185/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28114185/)!


Differential Revision: [D28114185](https://our.internmc.facebook.com/intern/diff/D28114185)

[ghstack-poisoned]

suo commented:
@suo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28114185).

suo committed:
Update on "tmp intern"




Differential Revision: [D28114185](https://our.internmc.facebook.com/intern/diff/D28114185/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28114185/)!


Differential Revision: [D28114185](https://our.internmc.facebook.com/intern/diff/D28114185)

[ghstack-poisoned]

suo committed:
Update on "tmp intern"




Differential Revision: [D28114185](https://our.internmc.facebook.com/intern/diff/D28114185/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28114185/)!


Differential Revision: [D28114185](https://our.internmc.facebook.com/intern/diff/D28114185)

[ghstack-poisoned]

suo commented:
@suo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28114185).

suo committed:
Update on "[package] Add an intern keyword"


Require that users be explicit about what they are going to be
interning. There are a lot of changes that are enabled by this. The new
overall scheme is:

PackageExporter maintains a dependency graph. Users can add to it,
either explicitly (by issuing a `save_*` call) or explicitly (through
dependency resolution). Users can also specify what action to take when
PackageExporter encounters a module (deny, intern, mock, extern).

Nothing (except pickles, tho that can be changed with a small amount
of work) is written to the zip archive until we are finalizing the
package. At that point, we consult the dependency graph and write out
the package exactly as it tells us to.

This accomplishes two things:
1. We can gather up *all* packaging errors instead of showing them one at a time.
2. We require that users be explicit about what's going in packages, which is a common request.

Differential Revision: [D28114185](https://our.internmc.facebook.com/intern/diff/D28114185/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28114185/)!


Differential Revision: [D28114185](https://our.internmc.facebook.com/intern/diff/D28114185)

[ghstack-poisoned]

suo commented:
@suo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28114185).

suo committed:
Update on "[package] Add an intern keyword"


Require that users be explicit about what they are going to be
interning. There are a lot of changes that are enabled by this. The new
overall scheme is:

PackageExporter maintains a dependency graph. Users can add to it,
either explicitly (by issuing a `save_*` call) or explicitly (through
dependency resolution). Users can also specify what action to take when
PackageExporter encounters a module (deny, intern, mock, extern).

Nothing (except pickles, tho that can be changed with a small amount
of work) is written to the zip archive until we are finalizing the
package. At that point, we consult the dependency graph and write out
the package exactly as it tells us to.

This accomplishes two things:
1. We can gather up *all* packaging errors instead of showing them one at a time.
2. We require that users be explicit about what's going in packages, which is a common request.

Differential Revision: [D28114185](https://our.internmc.facebook.com/intern/diff/D28114185/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28114185/)!


Differential Revision: [D28114185](https://our.internmc.facebook.com/intern/diff/D28114185)

[ghstack-poisoned]

SplitInfinity commented on torch/package/package_exporter.py:
```diff
@@ -429,6 +549,43 @@ def register_mock_hook(
         return handle
 
+    def register_intern_hook(
+        self, hook: Callable[["PackageExporter", str], None]
```
Perhaps a type alias would be useful for the hook type? It's repeated in three places now.

SplitInfinity commented on torch/package/package_exporter.py:
```diff
@@ -588,6 +709,112 @@ def _write(self, filename, str_or_bytes):
+
+            raise PackagingError(denied, broken, unhandled, include_filter)
+
+        for pattern, _, allow_empty in self.patterns:
+            if not allow_empty and pattern not in self.matched_patterns:
+                raise EmptyMatchError(
+                    f"Exporter did not match any modules to {pattern}, which was marked as allow_empty=False"
+                )
```
One of the goals of this PR is to aggregate and present all errors at once, but these `EmptyMatchErrors` are not aggregated and will still be presented one at a time. Could we aggregate these as well? That is, could we raise one exception with *all* unmatched patterns?

suo replied:
I'll file a follow-up issue for this (it's a little bit more involved to change)

SplitInfinity commented on torch/package/package_exporter.py:
```diff
@@ -588,6 +709,112 @@ def _write(self, filename, str_or_bytes):
+        broken = {}
+        unhandled = set()
+        for module_name, attrs in self.dependency_graph.nodes.items():
+            if attrs.get("action") == _ModuleProviderAction.DENY:
```
Thoughts on creating constants for these node attribute names? We're using `get` here, so any new code that is added that makes a type won't produce a `KeyError` and might take longer to debug than necessary.

suo replied:
yeah, I plan to actually factor attribute data into a real class and actually have some structure, but I want some practical experience on what that structure will look like first.

SplitInfinity commented on torch/package/package_exporter.py:
```diff
@@ -588,6 +709,112 @@ def _write(self, filename, str_or_bytes):
+            if attrs.get("action") is None:
+                unhandled.add(module_name)
+
+        if len(denied) != 0 or len(broken) != 0 or len(unhandled) != 0:
```
```suggestion
        if denied or broken or unhandled:
```

suo replied:
dammit, it *is* more readable!

SplitInfinity commented on torch/package/package_exporter.py:
```diff
@@ -588,6 +709,112 @@ def _write(self, filename, str_or_bytes):
+                source = attrs["source"]
+                self._write_source_string(module_name, source, is_package)
+
+            else:
+                raise AssertionError(
+                    f"Invalid action: {module_name}, {action}. Please report a bug to PyTorch."
+                )
```
I think it is unintuitive that nodes that matched some deny pattern are present in the graph with `action = DENY`, and that such graphs are expected to be detected in `_validate_dependency_graph`.

`extern`, `intern`, and `mock` all convey information about how the dependency graph should be executed (i.e. do you pull in the deps of a given node or not) but deny does not.

suo replied:
The reason for putting them in the graph is mostly so that we can continue on and present all errors to the user. We should never pass validation if there are denied modules in the graph though.

suo commented on test/package/test_dependency_api.py:
```diff
@@ -218,6 +207,76 @@ def test_allow_empty_with_error(self):
+        except PackagingError as e:
+            exception_raised = True
+            self.assertEqual(e.unhandled, set(["package_a", "package_a.subpackage"]))
+
+        self.assertTrue(exception_raised)
```
finally, a use for else after except

suo commented on test/package/test_dependency_api.py:
```diff
@@ -218,6 +207,76 @@ def test_allow_empty_with_error(self):
+            he.save_pickle("obj", "obj.pkl", obj2)
+
+    def test_broken_dependency(self):
+        """A unpackageable dependency should raise a BrokenDependencyError."""
```
oops left in by old change

suo commented on torch/package/package_exporter.py:
```diff
@@ -201,18 +271,36 @@ def save_source_string(
+
+        Arguments are otherwise the same as for :meth:`save_source_string`.
+        """
+        extension = "/__init__.py" if is_package else ".py"
+        filename = module_name.replace(".", "/") + extension
```
we are writing to the zip exporter, not the file system, which uses `/` as the separator.

suo commented on torch/package/package_exporter.py:
```diff
@@ -631,8 +851,10 @@ def _filename(self, package, resource):
     def _can_implicitly_extern(self, module_name: str):
-        return module_name == "torch" or (
-            module_name not in _DISALLOWED_MODULES and is_stdlib_module(module_name)
+        top_level_package_name = module_name.partition(".")[0]
```
wait why is this better lol

suo committed:
Update on "[package] Add an intern keyword"


Require that users be explicit about what they are going to be
interning. There are a lot of changes that are enabled by this. The new
overall scheme is:

PackageExporter maintains a dependency graph. Users can add to it,
either explicitly (by issuing a `save_*` call) or explicitly (through
dependency resolution). Users can also specify what action to take when
PackageExporter encounters a module (deny, intern, mock, extern).

Nothing (except pickles, tho that can be changed with a small amount
of work) is written to the zip archive until we are finalizing the
package. At that point, we consult the dependency graph and write out
the package exactly as it tells us to.

This accomplishes two things:
1. We can gather up *all* packaging errors instead of showing them one at a time.
2. We require that users be explicit about what's going in packages, which is a common request.

Differential Revision: [D28114185](https://our.internmc.facebook.com/intern/diff/D28114185/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28114185/)!


Differential Revision: [D28114185](https://our.internmc.facebook.com/intern/diff/D28114185)

[ghstack-poisoned]

suo committed:
Update on "[package] Add an intern keyword"


Require that users be explicit about what they are going to be
interning. There are a lot of changes that are enabled by this. The new
overall scheme is:

PackageExporter maintains a dependency graph. Users can add to it,
either explicitly (by issuing a `save_*` call) or explicitly (through
dependency resolution). Users can also specify what action to take when
PackageExporter encounters a module (deny, intern, mock, extern).

Nothing (except pickles, tho that can be changed with a small amount
of work) is written to the zip archive until we are finalizing the
package. At that point, we consult the dependency graph and write out
the package exactly as it tells us to.

This accomplishes two things:
1. We can gather up *all* packaging errors instead of showing them one at a time.
2. We require that users be explicit about what's going in packages, which is a common request.

Differential Revision: [D28114185](https://our.internmc.facebook.com/intern/diff/D28114185/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28114185/)!


Differential Revision: [D28114185](https://our.internmc.facebook.com/intern/diff/D28114185)

[ghstack-poisoned]

SplitInfinity commented on torch/package/package_exporter.py:
```diff
@@ -246,78 +338,101 @@ def _write_dep_graph(self, failing_module=None):
 
-        for i, (pattern, action, _) in enumerate(self.patterns):
+        for pattern, action, _ in self.patterns:
             if pattern.matches(module_name):
```
Is it possible for a module name to match multiple patterns? I think that would be user error, but this code will take the first rule that matches and ignore the others. Should we detect this case and provide a better error message?

E.g what if some module matches a generic intern glob as well as a more specific deny glob, but the intern happened to have been "declared" first?

suo replied:
It is intended behavior (and what currently happens) for patterns to match in order, with the first one winning. This allows people to specify catch-alls near the bottom. Our docs should make this clear though.

SplitInfinity reviewed (commented):
🚀

suo commented on torch/package/package_exporter.py:
```diff
@@ -588,6 +704,112 @@ def _write(self, filename, str_or_bytes):
+
+            if action == _ModuleProviderAction.EXTERN:
+                for hook in self._extern_hooks.values():
+                    hook(self, module_name)
```
we want hooks to be able to raise exceptions that terminate packaging though, that's the only use case rn (ExternNotAllowed). What would go in the finally clause?

SplitInfinity replied:
I left this comment before reviewing internal changes, I see that now

suo commented on torch/package/package_exporter.py:
```diff
@@ -588,6 +704,112 @@ def _write(self, filename, str_or_bytes):
+                # The node in the dependency graph contains metadata that tells us
+                # how to intern the module.
+                if "provided" not in attrs:
+                    raise AssertionError(
+                        f"Module was marked `intern` but not provided: {module_name}"
+                    )
```
assert statements don't execute in Python's "opt" (lol) mode. Other than that, I think there is no difference.

suo commented on torch/package/package_exporter.py:
```diff
@@ -246,78 +338,101 @@ def _write_dep_graph(self, failing_module=None):
+            )
+            return
+
+        self.dependency_graph.add_node(
+            module_name, is_package=is_package, source=source, provided=True
```
yeah really we want "update_node", but the networkx API doesn't have that (it's just add_node). Can I add as a followup? I don't want to add tests for the new digraph API to this PR if I can avoid.

SplitInfinity replied:
Ok

suo commented on torch/package/glob_group.py:
```diff
@@ -46,6 +46,9 @@ def __init__(
     def __str__(self):
         return self._dbg
 
+    def __repr__(self):
+        return self._dbg
```
Oh, this was a debugging change left in. I'll revert (although I agree with this comment and will probably do something like it in a followup)

suo commented on torch/package/package_exporter.py:
```diff
@@ -129,10 +205,8 @@ def __init__(
-            Tuple[Any, Callable[[str], None], bool]
-        ] = []  # 'any' is 're.Pattern' but breaks old mypy
-        self.matched_patterns: Set[int] = set()
+        self.patterns: List[Tuple[GlobGroup, _ModuleProviderAction, bool]] = []
+        self.matched_patterns: Set[GlobGroup] = set()
```
yeah, I can do the dictionary thing.
